### PR TITLE
Middleware page for documentation

### PR DESCRIPTION
**Pull request by EricWVGG**
Describes FileMiddleware, ErrorMiddleware, CORSMiddleware, and a link out to the Auth library.

A future version of this document might describe the creation of custom Middlewares. In the meantime, a link to the API is provided.

### CODE REVIEW

File: 3.0/docs/vapor/middleware.md [hunk 1]
```diff
@@ -1,3 +1,58 @@
 # Middleware
 
-Coming soon.
+This guide will introduce you to middleware. Middlewares are a type of code that operate “in the middle” of a request coming in and out of the Vapor software.
```
Comment (calebkleveter) [inline 3.0/docs/vapor/middleware.md]: This paragraph should probably be changed to something like:

> Middleware is a logic chain between the client and a Vapor route handler. It allows you to make operations on incoming requests before they get to the route handler and outgoing responses before they go to the client.

File: 3.0/docs/vapor/middleware.md [hunk 2]
```diff
@@ -1,3 +1,58 @@
...
+
+## Configuration, and ErrorMiddleware
+
+Middleware is registered in your `config.swift` file. ErrorMiddleware is a very common example; it will take a thrown error in your software and convert it to a legible HTTP response code.
```
Comment (calebkleveter) [inline 3.0/docs/vapor/middleware.md]: Middleware names should have back-tics (\`) around them so they are rendered as code: `ErrorMiddleware`

File: 3.0/docs/vapor/middleware.md [hunk 3]
```diff
@@ -1,3 +1,58 @@
...
+middlewares.use(FileMiddleware.self)
+// etc.
+services.register(middlewares)
+```
+
+It is common to run several middlewares in a single project. These middlewares are stacked up, and then registered together. The order in which middleware are listed can sometimes matter (see CORSMiddleware below).
+
+## FileMiddleware
+
+FileMiddleware enables the serving assets from the Public folder of your project to clients. You might include static files like stylesheets or bitmap images here.
```
Comment (calebkleveter) [inline 3.0/docs/vapor/middleware.md]: "the serving assets" -> "the serving of assets".
"your project to clients" -> "your project to the client".

**Comment by EricWVGG** [comment]
Changes made.

**Comment by EricWVGG** [comment]
@calebkleveter Changes are made. Any other requests?

File: 3.0/docs/vapor/middleware.md [hunk 4]
```diff
@@ -1,3 +1,58 @@
...
+    allowedMethods: [.GET, .POST, .PUT, .OPTIONS, .DELETE, .PATCH],
+    allowedHeaders: [.accept, .authorization, .contentType, .origin, .xRequestedWith, .userAgent, .accessControlAllowOrigin]
+)
+let corsMiddleware = CORSMiddleware(configuration: corsConfiguration)
+middlewares.use(corsMiddleware)
+middlewares.use(ErrorMiddleware.self)
+services.register(middlewares)
+```
+
+Given that thrown errors are immediately returned to the client, the CORSMiddleware must be listed _before_ the `ErrorMiddleware`; otherwise the HTTP error response will be returned without CORS headers, and cannot be read by the browser.
```
Comment (calebkleveter) [inline 3.0/docs/vapor/middleware.md]: Missed back-ticks round 'CORSMiddleware' here.

**Comment by calebkleveter** [comment]
Found one thing. Other then that, it looks good!

File: 3.0/docs/vapor/middleware.md (hunk 4 above)
Comment (EricWVGG) [reply 3.0/docs/vapor/middleware.md]: gotcha

**Comment by EricWVGG** [comment]
thumbs-up emoji

**Comment by penny-coin** [comment]
Hey @EricWVGG, you just merged a pull request, have a coin! 

You now have 1 coins.

**Pull request merged**